### PR TITLE
ENTERPRISE-857: utilities for encrypting/decrypting passwords in config.

### PR DESCRIPTION
**Pull request by jumarko**
New crypto namespace provides low-level functions encrypt-password and decrypt-password
with higher-level wrappers:
- `encrypt-user-password` for one-time encryption of plaintext password. This function is invoked
   via CLI interface (`--encrypt-password`) and the encrypted version can then be used in
   configuration
- `decrypt-secrets` for decrypting the encrypted passwords (jira & service) in the configuration

Encryption/decryption only happens when the `CODESCENE_JIRA_ENCRYPTION_KEY` is set.
Another optional env var `CODESCENE_JIRA_ENCRYPTION_NONCE` can be set to provide
a custom (but constant) "initialization vector" for encryption algorithm.

We use AES256 with GCM as an encryption algorithm.
For more details see the [`codescene-enterprise-pm-jira.crypto` documentation](src/codescene_enterprise_pm_jira/crypto.clj).

### CODE REVIEW

File: src/codescene_enterprise_pm_jira/crypto.clj [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                                         enc-key
+                                         (encryption-nonce)
+                                         encryption-algorithm)]
+    (bcodecs/bytes->str decrypted-bytes)))
+
+(defn encrypt-user-password
+  "Higher-level version of `encrypt-password` using/checking the value of the env var
+  and also providing error reporting."
+  [report-error-fn plaintext-password]
+  (if (string/blank? plaintext-password)
```
Comment (josf) [inline src/codescene_enterprise_pm_jira/crypto.clj]: Would it make sense to enforce a minimum password length?

File: README.md [hunk 2]
```diff
@@ -163,6 +163,83 @@ The configuration file path is resolved in the following order:
 If the configuration path doesn't point to a valid YAML file the service
 fails to start.
 
+### Encrypting passwords in configuration files
+
+If you don't want to accidently share the passwords by pushing your config file
```
[the remark this answers is not in the document]
Comment (jumarko) [reply README.md]: Fixed.

File: README.md [hunk 3]
```diff
@@ -163,6 +163,83 @@ The configuration file path is resolved in the following order:
...
+$ export CODESCENE_JIRA_ENCRYPTION_NONCE=1234567890ab
+
+# encrypt the service password
+$ java -jar target/codescene-enterprise-pm-jira.standalone.jar --encrypt-password
+2019-07-22 13:25:41.380:INFO::main: Logging initialized @4101ms
+Enter the password:
+
+Encrypted password:  ab18db35f9d25a4ebb6a868356f89b91d07ab7c0c9aa685d72620174eac0
+```
+Notice that the encrypted password is different from the case when the default nonce is used.
```
[the remark this answers is not in the document]
Comment (jumarko) [reply README.md]: Removed completely.

File: src/codescene_enterprise_pm_jira/crypto.clj [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+;; if this var is set then we expect encrypted passwords
+(def encryption-key-env-var "CODESCENE_JIRA_ENCRYPTION_KEY")
+
+;; Initialization vector (nonce) used for encryption
+;; Despite the usual recommendations it's constant (either hard-coded or set in env var)
+;; because we don't store passwords/nonces anywhere.
+;; Since we only encrypt a few secrets it shouldn't make a big difference.
+(def encryption-nonce-env-var "CODESCENE_JIRA_ENCRYPTION_NONCE")
+(def default-encryption-nonce "abcdefghijkl")
```
[the remark this answers is not in the document]
Comment (jumarko) [reply src/codescene_enterprise_pm_jira/crypto.clj]: Obsolete after using jasypt.

File: src/codescene_enterprise_pm_jira/crypto.clj [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  Note that we're not using CBC mode cipher since that was discontinued in TLS 1.3.
+  As per suggestions in https://stackoverflow.com/questions/1220751/how-to-choose-an-aes-encryption-mode-cbc-ecb-ctr-ocb-cfb
+  we use GCM instead.
+
+  About managing 'initialization vectors' (nonces):
+  - https://security.stackexchange.com/questions/17044/when-using-aes-and-cbc-is-it-necessary-to-keep-the-iv-secret"
+  (:require
+   [buddy.core
+    [codecs :as bcodecs]
+    [crypto :as bcrypto]
```
[the remark this answers is not in the document]
Comment (jumarko) [reply src/codescene_enterprise_pm_jira/crypto.clj]: Obsolete after using jasypt.

File: src/codescene_enterprise_pm_jira/crypto.clj (hunk 1 above)
Comment (jumarko) [reply src/codescene_enterprise_pm_jira/crypto.clj]: Yes, I think this may be a good idea.
8 chars?

**Pull request merged**